compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,369 @@
+    old_shape_end_idx++;
+    return true;
+  };
+
+  auto move_to_next_range = [&]() -> bool {
+    if (old_shape_end_idx > _shape.size())
+      return false;
+
+    old_shape_start_idx = old_shape_end_idx;
+    old_shape_end_idx++;
+    old_shape_product = _shape[old_shape_start_idx].value;
+    return true;
+  };
+
+  // Add tags from '_shape' to the 'new_shape'
+  uint32_t new_shape_idx = 0;
+  while (new_shape_idx < new_shape.size())
+  {
+    Dim &target_dim = new_shape[new_shape_idx];
+    if (target_dim.value == 1)
+    {
+      new_shape_idx++;
+      continue;
+    }
+
+    while (old_shape_product < target_dim.value)
+    {
+      if (expand_old_shape_range() == false)
+        break;
+    }
+
+    if (old_shape_product != target_dim.value)
+    {
+      _can_remove_transpsoe = false;
+      break;
+    }
+
+    // old_shape_product == target_dim.value
+    for (uint32_t idx = old_shape_start_idx; idx < old_shape_end_idx; idx++)
+    {
+      const auto &old_tags = _shape[idx].tags;
+      target_dim.tags.insert(target_dim.tags.end(), old_tags.begin(), old_tags.end());
+    }
+
+    new_shape_idx++;
+    move_to_next_range();
+  }
+  _shape = new_shape;
+  return;
+}
+
+void PatternSimulator::simulate()
+{
+  const luci::CircleNode *in_tensor = dynamic_cast<luci::CircleNode *>(_front_transpose->a());
+  init_shape(in_tensor);
+  simulate_transpose(_front_transpose);
+  simulate_reshape(_mid_reshape);
+  simulate_transpose(_back_transpose);
+}
+
+bool PatternSimulator::canRemoveTransposes() const
+{
+  if (_can_remove_transpsoe == false)
+    return false;
+
+  // check whether tags in _shape are incremental
+  uint8_t tag = _start_tag;
+  for (auto dim : _shape)
+  {
+    for (auto t : dim.tags)
+    {
+      if (t == tag)
+        tag++;
+      else
+        return false;
+    }
+  }
+  return true;
+}
+
+luci::CircleReshape *PatternSimulator::createAlternative(loco::Graph *graph)
+{
+  std::string pattern_name =
+    _front_transpose->name() + "/" + _mid_reshape->name() + "/" + _back_transpose->name();
+
+  auto shape_node = graph->nodes()->create<luci::CircleConst>();
+  {
+    shape_node->dtype(loco::DataType::S32);
+    shape_node->rank(1);
+    shape_node->dim(0).set(_back_transpose->rank());
+    shape_node->size<loco::DataType::S32>(_back_transpose->rank());
+    for (uint32_t i = 0; i < _back_transpose->rank(); i++)
+    {
+      shape_node->at<loco::DataType::S32>(i) = _back_transpose->dim(i).value();
+    }
+    shape_node->shape_status(luci::ShapeStatus::VALID);
+    shape_node->name(pattern_name + "/shape");
+  }
+
+  auto reshape_node = graph->nodes()->create<luci::CircleReshape>();
+  {
+    reshape_node->name(pattern_name);
+    reshape_node->tensor(_front_transpose->a());
+    reshape_node->shape(shape_node);
+    luci::add_origin(reshape_node, luci::get_origin(_mid_reshape));
+  }
+
+  return reshape_node;
+}
+
+bool remove_no_effect_reshape(luci::CircleTranspose *node)
```

Comment:
Ah, my mistake..
Sorry for confusion. 

```suggestion
bool remove_no_effect_transpose(luci::CircleTranspose *node)
```